Improve Backup/Restore/Migration docs, add v3 instructions

rocodes opened the pull request:
Include backup and restore instructions for v3 onion services. Add hardware migration instructions for v3 onion services and for cases where onion credentials may not be restored.

## Status

Work in progress

## Description of Changes

* Description:  Update Restore/Migration docs to include steps for v3 services. Slight edits to condense existing Backup/Restore copy. WIP: would appreciate review. 

* Fixes Issue #36


## Testing
manual review 

## Release 
none

## Checklist (Optional)

- [x] Doc linting (`make docs-lint`) passed locally
- [ ] Doc link linting (`make docs-linkcheck`) passed
- [x] You have previewed (`make docs`) docs at http://localhost:8000

rocodes commented on docs/backup_and_restore.rst:
```diff
@@ -123,67 +105,216 @@ Restoring
 
-Once SSH access to the servers has been established (or if using SSH over
-local network), Onion URLs for the *Source Interface* and *Journalist Interfaces*
-can be fetched using the installer:
```
@zenmonkeykstop: I did not use `./securedrop-admin install` to fetch onion URLs _to_ the Admin Workstation. I copied aths/auth-private files into place and ran tailsonfig, so I have provisionally removed this step. If I'm missing something, please let me know.

eloquence commented:
(Taking a first pass.)

eloquence commented on docs/admin.rst:
```diff
@@ -572,6 +574,11 @@ into your Admin Workstation, you should first perform the following troubleshoot
 
+#. **Ensure that Ansible is installed.** If Ansible is not installed, 
+   or ``./securedrop-admin tailsconfig`` fails, see
+   :doc:`configure_admin_workstation_post_install` for detailed setup instructions.
```
What motivates this addition? We already suggest running `./securedrop-admin tailsconfig` above; the command will instruct the admin to run `./securedrop-admin setup` if the virtual environment has not been initialized. The instructions you link to mainly ask the admin to run `./securedrop-admin tailsconfig`, so this feels somewhat circular.

rocodes replied:
Makes sense. I did pull it from the other troubleshooting steps and put it here as a last resort diagnostic (as in, if none of the above steps work, double-check that ansible is even installed). I wasn't sure how to put that. But if it reads circular we can remove it.

rocodes replied:
Removed this instruction in 151175d

eloquence commented on docs/backup_and_restore.rst:
```diff
@@ -1,12 +1,9 @@
-the system without changing Onion URLs, recreating journalist accounts,
+Maintaining regular backups helps guard against data 
+loss and hardware failure. Having a recent backup will allow you to redeploy
+SecureDrop without changing Onion URLs, recreating journalist accounts,
```
Yay for removing fluff sentences :)

eloquence commented on docs/backup_and_restore.rst:
```diff
@@ -123,67 +105,216 @@ Restoring
+   Then, edit the file called ``tor_v3_keys.json``, located in 
+   ``install_files/ansible-base`` to replace the two ``app_journalist`` and 
+   two ``app_ssh`` values in the file with their respective values 
+   that you preserved in the ``tor_v2_keys.json`` file from the previous installation.
```
`tor_v2_keys.json`->`tor_v3_keys.json`

rocodes replied:
yikes, fixed in 151175d

eloquence commented on docs/backup_and_restore.rst:
```diff
@@ -123,67 +105,216 @@ Restoring
 
 .. note:: The backup strategy used for SecureDrop versions prior to 0.3.7
           created encrypted archives with the extension ``.zip.gpg``.
           You can safely remove those files once you've created the ``.tar.gz``
```
^ SecureDrop 0.3.7 is a long time ago, I'd nominate this block for removal while we're overhauling this chapter.

rocodes replied:
Removed in 151175d

eloquence commented on docs/backup_and_restore.rst:
```diff
@@ -123,67 +105,216 @@ Restoring
+
+   .. code:: sh
+
+      ssh-keygen -f "/home/amnesia/.ssh/known_hosts" -R "app_ssh_url"
```
Because this command is essentially identical for both v2 and v3, I would suggest consolidating this a bit further: 1) Obtain the SSH service .onion address for the Application Server (instructions vary between v2 and v3); 2) Remove the old onion service form the list of known hosts (instructions are identical once you have the .onion address).

Regarding the sample command, it's not obvious to me at a glance that `app_ssh_url` is meant to be replaced by the user. Further up we use `<your_backup_file>` to indicate that the filename is to be substituted by the user, which has the problem of using reserved shell symbols `<` and `>` but is a pretty common pragma regardless. I'm open to other ideas, but I do think we want it to stand out a bit more.

rocodes replied:
Updated in 151175d - let me know if this wording works for you.

eloquence commented on docs/backup_and_restore.rst:
```diff
@@ -123,67 +105,216 @@ Restoring
----------
+#. **Preserve important credentials**: Before re-installing SecureDrop, you 
+   will need to preserve some important credential files from the old 
+   installation on the *Admin Workstation*. They should be temporarily saved
```
I would change "They should be .." to active voice, e.g., "Save these credentials in .."

eloquence commented on docs/backup_and_restore.rst:
```diff
@@ -123,67 +105,216 @@ Restoring
+   -  ``tor_v3_keys.json``
+
+
+   Once you have preserved those files, you may either delete or rename the 
```
I'm not sure we even want to enumerate deletion as an option at this stage, in case any contents of the directory are later needed for diagnostics. I would suggest just offering the ``securedrop.old`` step here.

eloquence commented on docs/backup_and_restore.rst:
```diff
@@ -123,67 +105,216 @@ Restoring
+
+   During configuration (``./securedrop-admin sdconfig``), set up the
+   new SecureDrop installation with the same onion 
+   services version(s) as your previous server configuration. Proceed through 
```
I think it may be desirable to spell this out a bit more to avoid confusion about what "onion service version(s)" means in this context. How about something like this:

> During the configuration step (`./securedrop-admin sdconfig`), you will be prompted to choose whether to enable v2 onion services, v3 onion services, or both. Make sure that your selection is the same as it was in the previous server configuration.

rocodes replied:
Resolved because we now instruct users to copy in `site-specific`

eloquence commented on docs/backup_and_restore.rst:
```diff
@@ -123,67 +105,216 @@ Restoring
+#. **Reconfigure Tor connections**: 
+   Run ``./securedrop-admin tailsconfig`` to update the *Admin Workstation*
+   to use the restored Onion URLs again. See :doc:`configure_admin_workstation_post_install`
+   for further instructions.
```
Perhaps add a recommendation to verify connectivity here?

rocodes replied:
Added in 151175d

eloquence commented on docs/backup_and_restore.rst:
```diff
@@ -123,67 +105,216 @@ Restoring
+Existing SecureDrop Installation <restore_data>`.
+
+If you are restoring data onto an instance whose onion URLs *have* changed,
+or you are performing a hardware migration, see :ref:`Migrating <migrating>`. 
```
The phrase "onion URLs have changed" seems ambiguous to me (relative to what?), and I think it could be difficult for an admin to choose between the "Migrating" pointer and the following "mismatched Tor config" pointer based on how this is currently worded. As I understand it, we want to cover two scenarios in this chapter:

- restoring data for an existing, working installation
- reinstalling and then restoring from a backup, either for hardware/platform migration or for disaster recovery

Depending on when the last backup was taken, both scenarios may require the use of `--preserve-tor-config` to avoid clobbering a working Tor config on the server. If that's accurate, I would suggest structuring the chapter into these two sections at a high level, and explaining the `--preserve-tor-config` behavior in context where appropriate.

rocodes replied:
Thank you for this feedback! In 151175d I've described the instructions as you indicate, and I've moved the section on `--preserve Tor config` to a section of "Additional Instructions" at the end. (Eventually I will include a small section in that Additionals section on rebuilding the monitor server, as per e.g. https://github.com/freedomofpress/securedrop/issues/5311, but for now the only thing in there is the `preserve Tor config` piece.)

eloquence commented:
Thanks @rocodes for filling in some very important detail for these instructions! Left some comments based on my initial review; the structural question may be easier to talk through synchronously esp. with @zenmonkeykstop to figure out what we want to address in the docs revision now, vs. defer to later.

rocodes commented:
@eloquence these comments are great, thank you. It was a little challenging figuring out how to clarify the steps here. First round of revisions coming later today.

zenmonkeykstop commented on docs/backup_and_restore.rst:
```diff
@@ -123,67 +105,213 @@ Restoring
+  - Restoring the backup to the new installation;
+  - Repairing Tor credentials for the new installation. 
+
+Detailed Migration Instructions
```
These instructions are a bit complex and branching - it might make sense to split them into 3 sections as follows:

1) v3-only - this should be simple: 
- copy `tor_v3_keys.json` to the fresh  clone of securedrop, 
- then proceed with a fresh install 
- and then restore from the backup

2) v2+v3 - actually, this will be the same as v3-only, as who cares about the v2 ssh setup, it's not used. You can copy app's `*ths` files as below if you want the `ssh app-legacy` and `ssh mon-legacy` commands to work
3) v2-only - this one is probably the most complex - you will need to 
- copy app's `*ths` files from old install, do a fresh install, 
- restore from backup (which will fail with an error that says to run tailsconfig)
- copy app's `*ths` files into place, 
- run tailsconfig, 
- then run restore again. 

The trick here is that you don't have to rebuild mon's config for v2, you can just set up a new ssh config. Also for the configs with a v3 element, all you should need is the `tor_v3_keys.json` file.

(Also I may be missing steps here, it's been a while!)

zenmonkeykstop replied:
Ooh also you'll need to copy keys from `securedrop.old` and you might as well copy `site-specific` while you're in there.

zenmonkeykstop replied:
Or you could skip the new clone and just move files you don't want out of the way....

eloquence replied:
So, what's the scenario in which we want to support them preserving v2 as part of a migration/reinstall? Should we just make it explicit that they should migrate to v3 while they're there? These docs didn't use to exist at all, if we add them now, it feels like we might as well make them v3-only.

zenmonkeykstop replied:
Hypothetically? Someone may have a downed v2 instance with an old backup available, and they may want to restore it before migrating. Or they have some other reason unique to them for going right up to the wire on v2. But adding a "Data-only migration to v3" option might not be a bad idea. something like:
---
If you want to restore the data from a backup on a fresh instance with new onion service addresses:

- complete the fresh install first
- copy the backup file FILENAME to install_files/ansible base
- run `./securedrop-admin restore --preserve-tor-config FILENAME`
---
would do the trick. (this is v2/v3 only but if we disable new v2 installs next release that's a strong hint.)

rocodes replied:
So the "data-only migration to v3" section is in here, more opaquely referred to as  "Restoring Data Without Restoring Tor Configuration", and is for the case of instances with a v2 backup file who want to recover data (ahem, if certain deadlines are missed). 

I think there's an argument for skipping the v2-only hardware migration instructions, and telling users to contact Support. As soon as we stop allowing v2 onion services on fresh installs, the v2 migration instructions won't work. However, since they are written up I'm including them for now anyone says otherwise, and we can remove them in a few weeks' time.

eloquence replied:
@rocodes: 
> However, since they are written up I'm including them for now 

Sounds reasonable; perhaps we can just add a blurb at the top of the section, e.g.:

> If you are migrating or restoring an installation using v2 onion services, please be aware that support for them will be removed from SecureDrop starting in March 2021. We encourage you to migrate to v3 onion services as part of any migration. To do so, reinstall SecureDrop with v3 onion services enabled and restore your data without overwriting the generated Tor configuration **[link to section]**. The instructions for users of v2 onion services below will be removed in a future release.

rocodes replied:
Added in note on depreciation 254b0d9, and also split out migration instructions by onion services version.

zenmonkeykstop reviewed (changes requested):
I *think* the migration instructions could be simpler. Either way I'd recommend restructuring them to treat the v2, v3, v2+3 scenarios separately.

rocodes commented on docs/backup_and_restore.rst:
```diff
@@ -1,28 +1,25 @@
+doing backups.
 
 .. tip:: Although it varies, the average throughput of an onion service is
          about 300 kB/s, or roughly 2 hours for 2GB. Plan your backup and
```
https://metrics.torproject.org/onionperf-throughput.html?start=2020-08-01&end=2021-01-21&server=onion I'm guestimating here, but I'll update it to ~ 90 mins for 2 GB, based on ~3 Mbps

rocodes commented:
I've addressed @zenmonkeykstop's minor edits (for typos etc) in d913113, and also taken a stab at the restructuring/splitting into sections via onion service. His workflow is slightly different than the one I used (and I think his is better), but I haven't tested each path yet so I still consider this WIP.

rocodes commented on docs/backup_and_restore.rst:
```diff
@@ -1,28 +1,25 @@
-encourage Journalists to delete regularly unneeded submissions from the *Journalist
-Interface*.
+
+Encouraing journalists to regularly delete older, unneeded submissions from 
```
oop. fixed & thank you.

eloquence commented:
Thanks @rocodes. I've taken another visual spin through, and it looks great to me. 

@rocodes and @zenmonkeykstop: What level of additional technical review would you recommend for these changes? My main concern is that stepping through a full migration, let alone all 3, will take significant time. Perhaps we should decouple this specific docs update from the 1.7.0 release timetable?

rocodes commented:
I would like to step through a full migration on Monday before asking for anyone else's review, and I think it's doable. I think it's reasonable to decouple this from 1.7.0 and land it even a day or two after release day.

rocodes commented:
@zenmonkeykstop: on the v3 scenario, I'm not having success with the workflow you suggested. Specifically, after the `restore`, I still need to reboot the app server and repair ssh creds, subbing in the original `app-ssh.auth_private`, similar to the v2 workflow. `tor_v3_keys.json` alone doesn't seem to do the trick for me. Let's connect tomorrow to make sure I'm not missing anything.

rocodes commented:
Here are the v3 steps that worked for me @zenmonkeykstop, could use your eye before I write it up.

* copy tor_v3_keys.json, pubkey files, and site-specific to the fresh clone of securedrop, and preserve `app-ssh.auth_private` file elsewhere (for now)
* proceed with a fresh install, run tailsconfig, then run `ssh app` and `ssh mon` 
* copy in the backup file, restore from the backup, which will work but then time out
* reboot the app server (or restart tor via console) 
* copy in `app-ssh.auth_private` file, rerun `tailsconfig`
* run `install`, then run `tailsconfig` again.

zenmonkeykstop commented on docs/backup_and_restore.rst:
```diff
@@ -123,67 +107,308 @@ Restoring
+   -  ``tor_v3_keys.json``
+   - All ``.asc`` files (these correspond to your *Submission Public Key*, your OSSEC alerts public key, and (if configured) your journalist alerts public key
+   - The file ``~/Persistent/securedrop/install_files/ansible-base/group-vars/all/site-specific``
+   - (optional) ``app-ssh-aths`` and ``mon-ssh-aths``, if you wish to reinstate the ``app-ssh-legacy`` and ``mon-ssh-legacy`` commands
```
mon isn't affected by the restore, so the monitor server after the restore will still have its  address from the fresh install. This means there's some divergence between how to handle v2 and v3 in this case. You can just retain the fresh install mon-ssh-aths in the v2 case, but you would need to update the original tor_v3_keys.json, adding in the mon-ssh keypair from the fresh install's tor_v3_keys.json, before running either tailsconfig (which will update the auth_private file) or install (which will update the client auth files on the server).

zenmonkeykstop commented on docs/backup_and_restore.rst:
```diff
@@ -123,67 +107,308 @@ Restoring
+#. **Delete temporary files**: Remove any temporary files or directories
+   you created as you performed this migration, including the 
+   ``securedrop.old`` directory.
+
```
There's one other restore/migrate case to consider based on changes in https://github.com/freedomofpress/securedrop/pull/5677 - namely the v2+v3 -> v3 one (previously didn't work but now unblocked)

rocodes replied:
See my comment- I have addressed this in a note in front of the v2+v3 instructions, indicating that migrating to Ubuntu 20.04 --> follow the v3-only instructions.

eloquence commented on docs/backup_and_restore.rst:
```diff
@@ -44,53 +50,31 @@ currently on the *Application Server*: log in over SSH and run
-          approximate size of the backup file (since the majority of the backup
-          archive is the stored submissions), and Tails' **Disks** utility to
-          see how much free space you have on your persistent volume.
+Run ``git describe --exact-match`` to verify that you have a tagged SecureDrop
```
I would suggest conservatism until after April 30, given that out-of-date workstations will not have this change when it's needed, and so the check should still be performed manually during migrations. Agree that the 0.12.0 reference should go though.

rocodes commented on docs/backup_and_restore.rst:
```diff
@@ -123,67 +107,308 @@ Restoring
+The instructions differ depending on which onion services version 
+your instance is using.
+
+* :ref:`Instances using v3 onion services <migrate_v3>`
```
fixed in 5c29677, and addressed all other comments in same commit

rocodes commented on docs/backup_and_restore.rst:
```diff
@@ -1,37 +1,43 @@
 
 You can use the following command to determine the volume of submissions
 currently on the *Application Server*: log in over SSH and run
 ``sudo du -sh /var/lib/securedrop/store``.
```
fixed in  5c29677 (& all such code-blocks also)

rocodes commented:
@zenmonkeykstop: I think I've addressed all your comments.  I also stepped through the v3 instructions again last night. 

- For v3 services, there is no mucking around with `tor_v3_keys.json` _if_ you copy it into place before running the clean install, as detailed in these instructions. 
- I've separated the common instructions into an `includes`. Honestly, the only thing that differs about the v2 and v3 instructions is the "hey, if you want, you can also save the old `app-ssh-aths` file so your legacy commands work." That will be irrelevant real soon anyways, so we will be able to cut this document in half pretty soon.
- To address the v2v3 xenial -> v3 focal scenario, I've made a note on the v2/v3 instructions that, when restoring onto Focal, the correct path is just to follow the v3 steps. Does that work for you?  I've also written up Focal migration instructions (separate PR, tk later today) that indicate which migration documents to follow (basically, the "v3 docs" or the "preserve-tor-config" docs).

zenmonkeykstop commented:
^^ one thing to note about patched-in changes below - the migration instructions include a tag verification/checkout step, which currently refers to 1.7.1. It should be 1.8.0,  this is updated by update_version.sh which will be run prior to pushing to stable.

rocodes commented:
Closing in favour of #168
